radix-engine/src/engine/call_frame.rs

```diff
@@ -568,6 +592,95 @@ where
 
         Ok((output, moving_buckets, moving_proofs, moving_vaults))
     }
+
+    fn take_values(
+        &mut self,
+        value_ids: &HashSet<StoredValueId>,
+    ) -> Result<Vec<StoredValue>, RuntimeError> {
+        let values = self.take_set(value_ids)?;
+        for value in &values {
+            if let StoredValue::KeyValueStore(_, store) = value {
+                for id in store.all_descendants() {
+                    self.refed_values.remove(&id);
+                }
+            }
+        }
+        Ok(values)
```

Comment:
Time to remove flat map of local vaults and kv stores, in favor of a tree of values?

```
Frame 
|- bucket_id => bucket
|- proof_id => proof
|- kv_store_id => kv_store 
             |- vault_id => vault
             |- kv_store_id => kv_store
                     |- vault_id => vault
                     |- kv_store_id => kv_store
```

All these vault id and kv store id should be `u32` instead of `Hash, u32`. To access a "local" vault, one must specify the full path; accessing local vault and globalized vault should be different APIs.

Another interesting thing we can potentially enforce is a lazy_map can only be globalized when the "local" children is empty.